pkg/cri/client/v1alpha2/client.go

```diff
@@ -0,0 +1,901 @@
+// Copyright Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha2
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+
+	v1 "k8s.io/cri-api/pkg/apis/runtime/v1"
+	api "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+)
+
+type Client interface {
+	v1.ImageServiceClient
+	v1.RuntimeServiceClient
+}
+
+type client struct {
+	logger.Logger
+	isc api.ImageServiceClient
+	rsc api.RuntimeServiceClient
+	rcc *grpc.ClientConn
+	icc *grpc.ClientConn
+}
+
+// Connect v2alpha1 RuntimeService and ImageService clients.
+func Connect(runtime, image *grpc.ClientConn) (Client, error) {
+	c := &client{
+		Logger: logger.Get("cri/client"),
+		rcc:    runtime,
+		icc:    image,
+	}
+
+	if c.rcc != nil {
+		c.Info("probing CRI v1alpha2 RuntimeService client...")
+		c.rsc = api.NewRuntimeServiceClient(c.rcc)
+		_, err := c.rsc.Version(context.Background(), &api.VersionRequest{})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if c.icc != nil {
+		c.Info("probing CRI v1alpha2 ImageService client...")
+		c.isc = api.NewImageServiceClient(c.icc)
+		_, err := c.isc.ListImages(context.Background(), &api.ListImagesRequest{})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return c, nil
+}
+
+// HasRuntimeService checks if the client is configured with runtime services.
+func (c *client) HasRuntimeService() bool {
+	return c.rcc != nil
+}
+
+// HasImageService checks if the client is configured with image services.
+func (c *client) HasImageService() bool {
+	return c.icc != nil
+}
+
+func (c *client) checkRuntimeService() error {
+	if c.rcc == nil {
+		return fmt.Errorf("no CRI v1alpha2 RuntimeService client")
+	}
+	return nil
+}
+
+func (c *client) checkImageService() error {
+	if c.rcc == nil {
+		return fmt.Errorf("no CRI v1alpha2 ImageService client")
+	}
+	return nil
+}
+
+func (c *client) Version(ctx context.Context, in *v1.VersionRequest, opts ...grpc.CallOption) (*v1.VersionResponse, error) {
+	if err := c.checkRuntimeService(); err != nil {
+		return nil, err
+	}
+
+	var (
+		req    api.VersionRequest
+		v1resp = &v1.VersionResponse{}
+	)
+
+	if err := v1ReqToAlpha(in, &req); err != nil {
+		return nil, err
+	}
+
+	resp, err := c.rsc.Version(ctx, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := alphaRespToV1(resp, v1resp); err != nil {
+		return nil, err
+	}
```

Comment:
nit/note: the "asymmetry" of the `req` and `v1resp` variables just caught my eye, ie. the other is pointer and the other is not. I think dropping the `var` block would improve the readability (and reduce LOC). WDYT?

```suggestion
	req := &api.VersionRequest{}
	if err := v1ReqToAlpha(in, req); err != nil {
		return nil, err
	}

	resp, err := c.rsc.Version(ctx, req)
	if err != nil {
		return nil, err
	}

	v1resp := &v1.VersionResponse{}
	if err := alphaRespToV1(resp, v1resp); err != nil {
		return nil, err
	}
```

Reply:
That asymmetry is a very good point, thanks for spotting it ! Rewrote it, not exactly as you suggested but instead using a var-block with two pointer assignments as my personal preference here.